### ISSUE TITLE
Feature request: Expand localfile log_format to include UNIX and TCP sockets, and pipe input

### ISSUE THREAD

**Issue by zbalkan** [wazuh/wazuh#15178]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.3.9 | Wazuh agent | Agent | N/A | OS version |

## Summary

Wazuh agent accepts local files and Windows event channel as the only log sources. Due to disk I/O, clients with high EPS may suffer from performance issues.

![image](https://user-images.githubusercontent.com/39981909/195999755-ed3e8470-3727-4b2f-97ed-d62298ac389f.png)

## Proposal

Currently, the Wazuh agent accepts logs from log sources configured via localfile configuraton's [log_format](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html#log-format) property.

However, the suggested way to forward applications which generate syslog format is to write them down in a file and read from there. The issue with this approach is the performance. Disk I/O bound way is less performant than using plain Unix or TCP sockets.

Rsyslog allows plain UDP/TCP output with [omfwd](https://www.rsyslog.com/doc/v8-stable/configuration/modules/omfwd.html), UNIX pipe output with [omuxsock](https://www.rsyslog.com/doc/v8-stable/configuration/modules/omuxsock.html), and pipe output with [ompipe](https://www.rsyslog.com/doc/v8-stable/configuration/modules/ompipe.html). It is relatively easy to consume those as streams and this way will provide higher performance by keeping eliminating disk I/O. 

Utilizing these Rsyslog output options as new forms of ingress, Wazuh agents can handle high volume of logs with less loss.

Utilizing pipes also can help Windows clients as Windows clients can use that option, too. However, the Windows pipe API is [completely different](https://macoy.me/blog/programming/Sockets) from UNIX pipe.

## Conclusion

Add the formats as ingress options to localfile:

- Local UDP/TCP syslog
- UNIX socket
- Unix (and probably Windows) pipe

**Comment by vikman90** [comment]
Hi @zbalkan, thanks for this interesting proposal!

This feature is similar to the [Fluent forwarder module](https://documentation.wazuh.com/current/user-manual/capabilities/fluent-forwarder.html), except that it would transmit the logs to the manager.

Regarding to the syslog socket, the manager provides a similar feature to ingest syslog messages: [Remote syslog](https://documentation.wazuh.com/current/user-manual/capabilities/log-data-collection/how-it-works.html#remote-syslog).

We will take into account your proposal for the roadmap. In the meantime, you might find remote syslog ingestion useful.

Best regards.

**Comment by zbalkan** [comment]
Hi @vikman90,

We discussed this issue with great staff from community support. The issue with sending the logs bypassing the agent is that the logs will be divided on the server side. User needs to check logs from agent ID X and agent ID 000 at the same time and will try to match those. This is a huge impact on the usability of Wazuh. As a SIEM, it is Wazuh which should be correlatibg the events but now we are separating the sources and try to match them manually. And it will not work as expected since it is not how Wazuh is designed.

Basically when I go to an agent dashboard, I will not be seeing the whole picture. So that is not an acceptable use case.

**Comment by vikman90** [comment]
Hi @zbalkan,

Gotcha, I think I've understood well your need and your proposal. So, we would expend the Logcollector's formats with:

- Syslog listener:
```xml
<log_format>syslog_tcp</log_format>
<log_format>syslog_udp</log_format>
```

- UNIX sockets:
```xml
<log_format>unix_tcp</log_format>
<log_format>unix_udp</log_format>
```
`<location>` must be a path to the socket to listen.

- Pipe
```xml
<log_format>pipe_read</log_format>
<log_format>pipe_write</log_format>
```
`<location>` must be a path to the pipe.

I'm not sure if the last configuration makes sense, but I wonder if any existing solution that provides logs may create a pipe to write into it or expects that the reader creates it.

Best regards.

**Comment by zbalkan** [comment]
Hi @vikman90,

Yes, that is what I meant. Since there is not a benchmark, I cannot provide which option is more performant. The preference will be decided by the user based on their setup. In general, pipes are more performant readers considering TCP or Unix sockets. And if the user disabled the firewall for any reason, they could just expose the TCP sockets. UNIX sockets and pipes are more secure as they are accessed locally. It will be the user's decision at the end. Yet, I am positive that users will prefer pipes due to simplicity.

On the other hand, I need to comment on your design on [Unix sockets](https://man7.org/linux/man-pages/man7/unix.7.html). They are different than IP sockets and they should be something like below:

```xml
  <!-- UNIX socket type: SOCK_STREAM -->
  <log_format>unix_stream</log_format>

  <!-- UNIX socket type: SOCK_DGRAM -->
  <log_format>unix_dgram</log_format>

  <!-- UNIX socket type: SOCK_SEQPACKET -->
  <log_format>unix_seq</log_format>
```

About pipe ingestion, rsyslog writes to an existing named pipe as you mentioned. But syslog-ng is able to create [its own](https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.37/administration-guide/47#TOPIC-1829088), and it is optional. To have the intersection and minimum viable feature that might work on, either Wazuh agent can create a named pipe by default, or user can use `mkfifo` and configure Wazuh agent to read from there. That is an architectural decision and Wazuh team needs to make the decision based on the product's design principles. So, I have no suggestion on that matter. As long as it is well documented, both ways are ok.

Yet, there is still a problem about using the `log_format` property. The ingested log format can be syslog, audit log or else. It defines the log location, not the format. So, my suggestion would be more like below:

```xml
  <!-- IP socket type: TCP -->
  <log_format>...</log_format>
  <location type="tcp">port</location>

  <!-- IP socket type: UDP -->
  <log_format>...</log_format>
  <location type="udp">port</location>

  <!-- UNIX socket type: SOCK_STREAM -->
  <log_format>...</log_format>
  <location type="unix_stream">socket path</location>

  <!-- UNIX socket type: SOCK_DGRAM -->
  <log_format>...</log_format>
  <location type="unix_dgram">socket path</location>

  <!-- UNIX socket type: SOCK_SEQPACKET -->
  <log_format>...</log_format>
  <location type="unix_seq">socket path</location>

  <!-- Named pipe / FIFO -->
  <log_format>...</log_format>
  <location type="pipe">pipe name</location>
```

P.S: I'm changing the issue title as `log_format` has no use here.

**Comment by vikman90** [comment]
@zbalkan thank you very much for your comment.

I've added this feature request to the global roadmap. We will keep you posted!

Best regards.

**Comment by allamiro** [comment]
I think this feature request is still very relevant, especially for remote-site and lightweight relay deployments.

A common Wazuh architecture today is:

network device → rsyslog/Logstash → local file → Wazuh agent → Wazuh manager

This works, but it means the Wazuh agent depends on an external service just to receive basic Syslog traffic. For many branch-office or small-site use cases, it would be useful if the agent could optionally act as a native UDP/TCP Syslog listener and forward those events to the manager without requiring rsyslog or Logstash to write intermediate files.

Rsyslog would still be the better choice for advanced use cases such as complex filtering, disk-assisted queues, TLS Syslog, transformations, and very high-volume ingestion. However, native agent-side Syslog input would simplify basic deployments, reduce disk I/O from file-based collection, and make the Wazuh agent more flexible as a lightweight remote collector.

It would also help preserve the agent-based collection model instead of forcing users to bypass the agent and send Syslog directly to the manager.

**Comment by allamiro** [comment]
@vikman90  @zbalkan I'm implementing the agent-side ingress described here in phases against `wazuh/wazuh-agent` (6.0), keeping each piece small and independently reviewable. The 6.0 agent uses YAML configuration, so the listener definitions adapt the proposal in this thread to the current `logcollector` config style.

**Phase 1 — UDP/TCP syslog listeners:** wazuh/wazuh-agent#877 (open).
Native agent-side UDP/TCP syslog listeners in Logcollector, with support for multiple listeners, configuration validation (protocol, bind address, port, duplicate detection), clean startup/shutdown, and unit tests. Events enter the normal Logcollector pipeline and stay associated with the receiving agent.

**Phase 2 — UNIX domain sockets (`unix_stream` / `unix_dgram`):** implemented and ready.
Reuses the same reader pattern, plus stale socket-file cleanup before bind and `sun_path` length validation. I'll open it as a separate PR once #877 has been reviewed/merged, so the diff stays focused on the new code.

**Phase 3 — pipes / FIFO:** planned next, following the same pattern.

Each PR references this issue; the final one (pipes) will close it once the full set has landed. Happy to adjust the approach (for example, combining phases) based on maintainer preference.
